src/catalog/src/durable/impls/persist.rs

```diff
@@ -119,19 +119,44 @@ impl PersistHandle {
         bootstrap_args: &BootstrapArgs,
         deploy_generation: Option<u64>,
     ) -> Result<Box<dyn DurableCatalogState>, CatalogError> {
-        let (is_initialized, upper) = self.is_initialized_inner().await;
+        let mut initialized_status = self.is_initialized_inner().await;
 
-        if !matches!(mode, Mode::Writable) && !is_initialized {
+        if !matches!(mode, Mode::Writable) && !initialized_status.is_initialized() {
             return Err(CatalogError::Durable(DurableCatalogError::NotWritable(
                 format!("catalog tables do not exist; will not create in {mode:?} mode"),
             )));
         }
 
-        let user_version = self.get_user_version(upper).await;
+        // Listen cannot read any writes that happen at `Timestamp::minimum()`. So we initialize
+        // the shard with an empty write at `Timestamp::minimum()`.
+        if matches!(initialized_status, InitializedStatus::ShardUninitialized) {
+            let upper = initialized_status.upper();
+            let next_upper = upper.step_forward();
+            self.compare_and_append(Vec::new(), upper, next_upper)
+                .await?;
+            initialized_status = InitializedStatus::CatalogUninitialized;
+            soft_assert_eq!(next_upper, initialized_status.clone().upper());
+        }
```

Comment:
This approach ended up being much simpler than doing something like:

    Initialize shard with initial catalog state at Timestamp::minimum().
    Get snapshot and start listening.

Initializing the shard with state would either need to take a different write path than all other writes (which is to use a transaction) or I would have had to do some serious refactoring.

Reply:
So what I was thinking was that, up in `PersistHandle::new`, right after getting the write_handle, you could do:

```
const SENTINEL_UNINITIALIZED: T = T::minimum().step_forward();
let (write_handle, read_handle) = persist_client
...
let () = write_handle.append(&[], Antichain::from_elem(T::minimum()), Antichain::from_elem(SENTINEL_UNINITIALIZED)).await.expect("WIP").expect("WIP");
```

That append call will do an empty write if the upper is T::minimum and nothing otherwise. Then you can use `upper.elements() == &[SENTINEL_UNINITIALIZED]` as the check if it's initialized or not. Maybe that lets you go back to two states instead of having to deal with three?

Reply:
Just pushed an update with this, I think it looks a bit cleaner.